providers/apis/defi/raydium/multi_client.go

```diff
@@ -0,0 +1,116 @@
+package raydium
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"go.uber.org/zap"
+
+	oracleconfig "github.com/skip-mev/slinky/oracle/config"
+)
+
+// MultiJSONRPCClient is an implementation of the SolanaJSONRPCClient interface that delegates
+// requests to multiple underlying clients, and aggregates over all provided responses.
+type MultiJSONRPCClient struct {
+	// underlying clients
+	clients []SolanaJSONRPCClient
+
+	// logger
+	logger *zap.Logger
+}
+
+func NewMultiJSONRPCClient(clients []SolanaJSONRPCClient, logger *zap.Logger) *MultiJSONRPCClient {
+	return &MultiJSONRPCClient{
+		clients: clients,
+		logger:  logger,
+	}
+}
+
+// NewMultiJSONRPCClientFromEndpoints creates a new MultiJSONRPCClient from a list of endpoints.
+func NewMultiJSONRPCClientFromEndpoints(endpoints []oracleconfig.Endpoint, logger *zap.Logger) *MultiJSONRPCClient {
+	clients := make([]SolanaJSONRPCClient, len(endpoints))
+	for i := range endpoints {
+		client := rpc.New(endpoints[i].URL)
+		clients[i] = client
+	}
+	return NewMultiJSONRPCClient(clients, logger)
+}
+
+// GetMultipleAccountsWithOpts delegates the request to all underlying clients and applies a filter
+// to the responses.
+func (c *MultiJSONRPCClient) GetMultipleAccountsWithOpts(
+	ctx context.Context,
+	accounts []solana.PublicKey,
+	opts *rpc.GetMultipleAccountsOpts,
+) (*rpc.GetMultipleAccountsResult, error) {
+	// Create a channel to receive the responses from the underlying clients
+	responsesCh := make(chan *rpc.GetMultipleAccountsResult, len(c.clients))
+
+	// spawn a goroutine for each client to fetch the accounts
+	var wg sync.WaitGroup
+	wg.Add(len(c.clients))
+
+	for i := range c.clients {
+		go func(client SolanaJSONRPCClient) {
+			defer wg.Done()
+			resp, err := client.GetMultipleAccountsWithOpts(ctx, accounts, opts)
+			if err != nil {
+				c.logger.Error("failed to fetch accounts", zap.Error(err))
+				return
+			}
+			responsesCh <- resp
+		}(c.clients[i])
+	}
+
+	// close the channel once all responses are received, or the context is cancelled
+	go func() {
+		select {
+		case <-ctx.Done():
+			c.logger.Error("context cancelled")
+		case <-channelForWaitGroup(&wg):
+		}
+		close(responsesCh)
+	}()
+
+	responses := make([]*rpc.GetMultipleAccountsResult, 0)
```

Comment:
```suggestion
	responses := make([]*rpc.GetMultipleAccountsResult, len(responseCh))
```

we can preallocate this and not use `append` here

Reply:
If there are failed responses / the channel is closed before all respond, we'll have nil entries?

Reply:
I can set a capacity for the slice?